pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,348 @@
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te, nil)
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	// Only register if there is a trace in the context and it is set to verbose
+	// logging.
+	span := tracing.SpanFromContext(baseCtx.Context())
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return
+	}
+
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEventf(baseCtx.Context(), 2, "skip registering raft proposal without index: %v", ent)
+		return
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntry to not leak a registered permit.
+	if !r.maybeRegister() {
+		log.VEvent(baseCtx.Context(), 2, "too many active raft traces, skipping")
+		return
+	}
+
+	// Grab the trace and span id now because the baseCtx.Context may be
+	// nil'ed later.
+	tv := r.storeEntry(
+		kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+		baseCtx,
+	)
+	tv.logf(1, "registering local trace %s", tv)
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces on the store.
+	if r.numRegistered.Load() == 0 {
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// FlushAll will unregister all the currently active traces. It is safe to call
+// multiple times, but should always be called when the replica is destroyed.
+func (r *RaftTracer) FlushAll() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.m.Delete(index)
+		r.numRegistered.Add(-1)
+		t.logf(2, "cleanup log index %d during close", index)
+		return true
+	})
+}
+
+func peer(p raftpb.PeerID) redact.SafeString {
+	return redact.SafeString(raft.DescribeTarget(p))
+}
+
+// traceIfCovered will log the message if it touches any of the registered trace
+// points. Additionally it returns any saved contexts by index for sending to
+// remote nodes. This typically applies to messages that the leader sends to the
+// followers.
```

Comment:
This applies both the messages that the leader sends to followers, and all replicas send to their local storage.